File: nxflix-agents-ts/src/models/pipeline-config.ts
import { z } from 'zod';

/**
 * Script generation provider options.
 */
export const ScriptProvider = z.enum(['claude', 'openai', 'gemini']);
export type ScriptProvider = z.infer<typeof ScriptProvider>;

/**
 * Text-to-speech provider options.
 */
export const TTSProviderType = z.enum(['elevenlabs', 'openai', 'google', 'azure']);
export type TTSProviderType = z.infer<typeof TTSProviderType>;

/**
 * Image generation provider options.
 */
export const ImageProvider = z.enum(['dalle', 'gemini', 'stable_diffusion', 'static']);
export type ImageProvider = z.infer<typeof ImageProvider>;

/**
 * Video composition provider options.
 */
export const VideoProvider = z.enum(['ffmpeg', 'gemini_veo', 'runway', 'pika']);
export type VideoProvider = z.infer<typeof VideoProvider>;

/**
 * Video resolution options.
 */
export const VideoResolution = z.enum(['720p', '1080p', '4k']);
export type VideoResolution = z.infer<typeof VideoResolution>;

/**
 * Video output format options.
 */
export const VideoFormat = z.enum(['mp4', 'webm']);
export type VideoFormat = z.infer<typeof VideoFormat>;

/**
 * TTS provider-specific settings.
 */
export const TTSSettings = z.object({
  voice: z.string().optional(),
  speed: z.number().min(0.5).max(2.0).default(1.0),
  pitch: z.number().min(-20).max(20).default(0),
});
export type TTSSettings = z.infer<typeof TTSSettings>;

/**
 * Image provider-specific settings.
 */
export const ImageSettings = z.object({
  style: z.string().optional(),
  generateBackground: z.boolean().default(false),
  generateCharacter: z.boolean().default(false),
});
export type ImageSettings = z.infer<typeof ImageSettings>;

/**
 * Video composition settings.
 */
export const VideoSettings = z.object({
  resolution: VideoResolution.default('1080p'),
  fps: z.number().int().min(15).max(60).default(30),
  format: VideoFormat.default('mp4'),
  codec: z.enum(['h264', 'h265', 'vp8', 'vp9']).default('h264'),
});
export type VideoSettings = z.infer<typeof VideoSettings>;

/**
 * Complete pipeline configuration for video generation.
 * Allows users to select which AI provider to use for each step.
 */
export const PipelineConfig = z.object({
  scriptProvider: ScriptProvider.default('claude'),
  ttsProvider: TTSProviderType.default('openai'),
  imageProvider: ImageProvider.default('static'),
  videoProvider: VideoProvider.default('ffmpeg'),
  ttsSettings: TTSSettings.default({}),
  imageSettings: ImageSettings.default({}),
  videoSettings: VideoSettings.default({}),
});
export type PipelineConfig = z.infer<typeof PipelineConfig>;

/**
 * Provider availability status.
 */
export const ProviderStatus = z.object({
  id: z.string(),
  name: z.string(),
  available: z.boolean(),
  reason: z.string().optional(),
});
export type ProviderStatus = z.infer<typeof ProviderStatus>;

/**
 * Voice option for TTS.
 */
export const VoiceOption = z.object({
  id: z.string(),
  name: z.string(),
  gender: z.string(),
  language: z.string().default('ja-JP'),
  provider: TTSProviderType,
  description: z.string().optional(),
});
export type VoiceOption = z.infer<typeof VoiceOption>;

/**
 * Response containing all available providers and their status.
 */
export const ProvidersResponse = z.object({
  script: z.array(ProviderStatus),
  tts: z.array(
    ProviderStatus.extend({
      voices: z.array(VoiceOption).optional(),
    })
  ),
  image: z.array(ProviderStatus),
  video: z.array(ProviderStatus),
});
export type ProvidersResponse = z.infer<typeof ProvidersResponse>;

/**
 * Default pipeline configuration.
 * Uses the most reliable providers with reasonable defaults.
 */
export const DEFAULT_PIPELINE_CONFIG: PipelineConfig = {
  scriptProvider: 'claude',
  ttsProvider: 'openai',
  imageProvider: 'static',
  videoProvider: 'ffmpeg',
  ttsSettings: {
    speed: 1.0,
    pitch: 0,
  },
  imageSettings: {
    generateBackground: false,
    generateCharacter: false,
  },
  videoSettings: {
    resolution: '1080p',
    fps: 30,
    format: 'mp4',
    codec: 'h264',
  },
};

/**
 * Video provider fallback order.
 * If the primary provider fails, try the next one.
 */
export const VIDEO_PROVIDER_FALLBACK_ORDER: VideoProvider[] = [
  'gemini_veo',
  'runway',
  'pika',
  'ffmpeg',
];

File: nxflix-agents-python/src/nxflix_agents/models/pipeline_config.py
"""
Pipeline configuration models for video generation.
Allows users to select which AI provider to use for each step.
"""

from enum import Enum
from typing import Optional, List
from pydantic import BaseModel, Field


class ScriptProvider(str, Enum):
    """Script generation provider options."""
    CLAUDE = "claude"
    OPENAI = "openai"
    GEMINI = "gemini"


class TTSProviderType(str, Enum):
    """Text-to-speech provider options."""
    ELEVENLABS = "elevenlabs"
    OPENAI = "openai"
    GOOGLE = "google"
    AZURE = "azure"


class ImageProvider(str, Enum):
    """Image generation provider options."""
    DALLE = "dalle"
    GEMINI = "gemini"
    STABLE_DIFFUSION = "stable_diffusion"
    STATIC = "static"


class VideoProvider(str, Enum):
    """Video composition provider options."""
    FFMPEG = "ffmpeg"
    GEMINI_VEO = "gemini_veo"
    RUNWAY = "runway"
    PIKA = "pika"


class VideoResolution(str, Enum):
    """Video resolution options."""
    HD_720P = "720p"
    FULL_HD_1080P = "1080p"
    UHD_4K = "4k"


class VideoFormat(str, Enum):
    """Video output format options."""
    MP4 = "mp4"
    WEBM = "webm"


class TTSSettings(BaseModel):
    """TTS provider-specific settings."""
    voice: Optional[str] = None
    speed: float = Field(default=1.0, ge=0.5, le=2.0)
    pitch: float = Field(default=0, ge=-20, le=20)


class ImageSettings(BaseModel):
    """Image provider-specific settings."""
    style: Optional[str] = None
    generate_background: bool = False
    generate_character: bool = False


class VideoSettings(BaseModel):
    """Video composition settings."""
    resolution: VideoResolution = VideoResolution.FULL_HD_1080P
    fps: int = Field(default=30, ge=15, le=60)
    format: VideoFormat = VideoFormat.MP4
    codec: str = "h264"


class PipelineConfig(BaseModel):
    """Complete pipeline configuration for video generation."""
    script_provider: ScriptProvider = ScriptProvider.CLAUDE
    tts_provider: TTSProviderType = TTSProviderType.OPENAI
    image_provider: ImageProvider = ImageProvider.STATIC
    video_provider: VideoProvider = VideoProvider.FFMPEG
    tts_settings: TTSSettings = Field(default_factory=TTSSettings)
    image_settings: ImageSettings = Field(default_factory=ImageSettings)
    video_settings: VideoSettings = Field(default_factory=VideoSettings)


class ProviderStatus(BaseModel):
    """Provider availability status."""
    id: str
    name: str
    available: bool
    reason: Optional[str] = None


class VoiceOption(BaseModel):
    """Voice option for TTS."""
    id: str
    name: str
    gender: str
    language: str = "ja-JP"
    provider: TTSProviderType
    description: Optional[str] = None


class TTSProviderStatus(ProviderStatus):
    """TTS provider status with available voices."""
    voices: Optional[List[VoiceOption]] = None


class ProvidersResponse(BaseModel):
    """Response containing all available providers and their status."""
    script: List[ProviderStatus]
    tts: List[TTSProviderStatus]
    image: List[ProviderStatus]
    video: List[ProviderStatus]


# Default configuration
DEFAULT_PIPELINE_CONFIG = PipelineConfig(
    script_provider=ScriptProvider.CLAUDE,
    tts_provider=TTSProviderType.OPENAI,
    image_provider=ImageProvider.STATIC,
    video_provider=VideoProvider.FFMPEG,
)

# Video provider fallback order
VIDEO_PROVIDER_FALLBACK_ORDER = [
    VideoProvider.GEMINI_VEO,
    VideoProvider.RUNWAY,
    VideoProvider.PIKA,
    VideoProvider.FFMPEG,
]
